pandas/io/parsers/base_parser.py

```diff
@@ -930,7 +937,23 @@ def _check_data_length(
                 stacklevel=find_stack_level(),
             )
 
-    def _evaluate_usecols(self, usecols, names):
+    @overload
+    def _evaluate_usecols(
+        self, usecols: set[int] | Callable[[Hashable], int], names: Sequence[Hashable]
```

Comment:
can the callable return a bool?

Reply:
No, the return values are indices

Reply:
sorry should have been more clear.

I'm curious as to why usecols is a callable with an `int` return type.

`ParserBase.__init__` is not typed, so it is not clear to me why this is `int`. `bool` is type compatible here and surely any truthy value is valid. So the return type of the usecols callable is whatever the public api accepts. This is any object that can be truthy?

Reply:
ah got you, yeah you are correct, this has to return a bool

Reply:
in the docs for say read_csv https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.read_csv.html

> If callable, the callable function will be evaluated against the column names, returning names where the callable function evaluates to True.

So doesn't explicitly say it should be `bool`, but why has `int` been chosen?

Reply:
Because I minsinterpreted the return value of the callable as the return value of _evaluate_use_cols

Reply:
sure. we do use `bool` in many places in the public api where the api probably accepts any object that can be truthy. This may become an issue for users when the types are public and users start getting false positives when type checking.

so changing to bool is fine for now.

Reply:
Technically this could return anything, which evaluates to True/False. But since the docs say that this has to evaluate to True, I think we can type it like this?

Reply:
Based on the doc-string (I didn't look at the code), it should probably be `Callable[[Hashable], object]`.

Reply:
Changed to object